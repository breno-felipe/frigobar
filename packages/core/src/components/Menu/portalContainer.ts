/**
 * Creates DOM element to be used as React root.
 */
function createRootElement(id: string) {
  const rootContainer = document.createElement('div');
  rootContainer.setAttribute('id', id);

  return rootContainer;
}

/**
 * Appends element as last child of body.
 */
function addRootElement(rootElem: HTMLElement) {
  document.body.appendChild(rootElem);
}

/**
 * Creates or return the current container DOM element to be the portal root.
 */
const portalContainer = (id: string): HTMLDivElement => {
  const rootElement = createRootElement(id);

  const parent = document.querySelector(`#${id}`);

  if (!parent) {
    addRootElement(rootElement);
  }

  return rootElement;
};

export default portalContainer;
