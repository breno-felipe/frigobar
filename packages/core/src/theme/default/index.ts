import { DefaultTheme } from 'styled-components';
import colors from './colors';
import spacings from './spacings';
import borders from './borders';
import radius from './radius';
import baseFontSize from './fontSize';

const theme: DefaultTheme = {
  colors,
  spacings,
  borders: borders,
  radius,
  baseFontSize,
};

export default theme;
