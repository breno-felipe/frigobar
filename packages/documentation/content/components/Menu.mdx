---
title: Menu
menu: components
---

Menu component

```jsx react-live=true noInline
render(() => {
  const buttonRef = useRef(null);
  const [open, toggleOpen] = useState(false);

  return (
    <div>
      <Menu
        anchorElement={buttonRef}
        open={open}
        handleClickAway={() => {
          toggleOpen(false);
        }}
      >
        <Menu.Item>Item 1</Menu.Item>
        <Menu.Item>Item 2</Menu.Item>
      </Menu>
      <Button ref={buttonRef} onClick={() => toggleOpen(true)}>
        Open menu
      </Button>
    </div>
  );
});
```
