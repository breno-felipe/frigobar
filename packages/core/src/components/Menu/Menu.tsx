import React, { useState, useEffect, useRef, useCallback } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import { useFade } from '@frigobar/animation';

import portalContainer from './portalContainer';

const PORTAL_CONTAINER_NAME = 'frigobar-menu';

const List = styled.ul<{ top: number; left: number }>(
  ({ top, left, theme: { colors, radius } }) => `
    position: absolute;
    top: ${top}px;
    left: ${left}px;
    z-index: 999;

    margin: 0;
    padding: 0;

    background-color: ${colors.white};
    border-radius: ${radius[1]}px;
    box-shadow: 1px 1px 3px rgba(0, 0, 0, 0.3);

    list-style: none;
    overflow: hidden;

    li {
      margin: 0;
      padding: 0;
    }
  `,
);

const Item = styled.a(
  ({ theme: { spacings, colors } }) => `
    display: block;
    padding: ${spacings.xsmall}px;
      
    color: inherit;

    cursor: pointer;
    text-decoration: none;

    &:hover {
      background-color: ${colors.primary[200]};
    }
  `,
);

interface MenuProps {
  children: React.ReactNode;
  /** a react component ref to be the anchor of the menu items */
  anchorElement: React.RefObject<HTMLElement>;
  /** whether the menu is visible or not */
  open: boolean;
  /** function called when user clicks outside the menu */
  handleClickAway?: (event: MouseEvent) => void;
  /** duration of the fade in/out animation */
  fadeDuration?: number;
}

const Menu = ({
  children,
  anchorElement,
  open,
  handleClickAway = () => null,
  fadeDuration,
  ...props
}: MenuProps): JSX.Element | null => {
  const [mounted, setMounted] = useState(false);
  const [anchorPosition, setAnchorPosition] = useState<{
    top: number;
    left: number;
  }>({ top: 0, left: 0 });
  const [{ animation, state }, toggleFade] = useFade({
    duration: fadeDuration,
  });

  const safeHandleClickAway = useCallback(event => handleClickAway(event), [
    handleClickAway,
  ]);

  const menuRef = useRef<HTMLUListElement>(null);

  const clickAway = useCallback(
    event => {
      if (
        !menuRef.current?.contains(event.target) &&
        anchorElement.current &&
        !anchorElement.current.contains(event.target)
      ) {
        safeHandleClickAway(event);
      }
    },
    [anchorElement, safeHandleClickAway],
  );

  useEffect(() => {
    setMounted(true);
  }, []);

  useEffect(() => {
    toggleFade(open);
  }, [open, toggleFade]);

  useEffect(() => {
    if (open) {
      window.addEventListener('click', clickAway);
    } else {
      window.removeEventListener('click', clickAway);
    }

    return () => {
      window.removeEventListener('click', clickAway);
    };
  }, [clickAway, open]);

  useEffect(() => {
    if (anchorElement.current) {
      if (open) {
        const {
          top,
          left,
          height,
        } = anchorElement.current?.getBoundingClientRect();

        setAnchorPosition({ top: top + height - 3, left: left + 3 });
      }
    }
  }, [anchorElement, open]);

  if (mounted) {
    portalContainer(PORTAL_CONTAINER_NAME);
  }

  return mounted
    ? createPortal(
        state ? (
          <List
            css={`
              animation: ${animation};
            `}
            ref={menuRef}
            top={anchorPosition.top}
            left={anchorPosition.left}
            {...props}
          >
            {React.Children.map(children, child => (
              <li>{child}</li>
            ))}
          </List>
        ) : null,
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        document.querySelector('#frigobar-menu')!,
      )
    : null;
};

List.displayName = 'Menu';
Item.displayName = 'Menu.Item';

Menu.Item = Item;

Menu.defaultProps = {
  fadeDuration: 300,
  open: false,
};

export default Menu;
