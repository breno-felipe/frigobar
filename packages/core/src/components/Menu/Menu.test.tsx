import React, { useState, useRef } from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';

import theme from '../../theme';

import Menu from '.';

const Component = ({ openned }: { openned?: boolean }) => {
  const [open, setOpen] = useState(openned);
  const buttonRef = useRef(null);

  return (
    <ThemeProvider theme={theme}>
      <div>
        <button onClick={() => setOpen(!open)} ref={buttonRef} type="button">
          trigger
        </button>
        <Menu anchorElement={buttonRef} open={open}>
          <Menu.Item>Item 1</Menu.Item>
          <Menu.Item>Item 2</Menu.Item>
        </Menu>
      </div>
      <div id="frigobar-menu"></div>
    </ThemeProvider>
  );
};

describe('<Menu />', () => {
  describe('snapshots', () => {
    it('should match', () => {
      const { container } = render(<Component openned />);
      expect(container).toMatchSnapshot();
    });
  });
});
