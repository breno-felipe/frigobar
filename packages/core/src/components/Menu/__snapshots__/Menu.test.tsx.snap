// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> snapshots should match 1`] = `
.c0 {
  position: absolute;
  top: -3px;
  left: 3px;
  z-index: 999;
  margin: 0;
  padding: 0;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: 1px 1px 3px rgba(0,0,0,0.3);
  list-style: none;
  overflow: hidden;
}

.c0 li {
  margin: 0;
  padding: 0;
}

.c1 {
  -webkit-animation: lbWRkT 300ms forwards;
  animation: lbWRkT 300ms forwards;
}

.c2 {
  display: block;
  padding: 8px;
  color: inherit;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:hover {
  background-color: #80c6cb;
}

<div>
  <div>
    <button
      type="button"
    >
      trigger
    </button>
  </div>
  <div
    id="frigobar-menu"
  >
    <ul
      class="c0 c1"
    >
      <li>
        <a
          class="c2"
        >
          Item 1
        </a>
      </li>
      <li>
        <a
          class="c2"
        >
          Item 2
        </a>
      </li>
    </ul>
  </div>
</div>
`;
